designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+    parser: require("babel-eslint"),
+    processor: require("eslint-plugin-markdown).processors.markdown
+};
+```
+
+In both cases, users now must pass a direct object reference. This has the benefit of using the builtin Node.js module resolution system or allowing users to specify their own. There is never of question of where the modules will be resolved from.
+
+#### Applying an Environment
+
+Unlike with `.eslintrc` files, there is no `env` key in `.eslint.js`. Users can mimmick the behavior of `env` by assigning directly to the `globals` key:
+
+```js
+const globals = require("globals");
+
+exports.config = {
+    globals: {
+        MyGlobal: true,
+        ...globals.browser
+    }
+};
+```
+
+This effectively duplicates the use of `env: { browser: true }` in ESLint.
+
+**Note:** This would allow us to stop environments in ESLint. We could just tell people to use `globals` in their config and allow them to specify which version of `globals` they want to use.
+
+#### Extending Another Config
+
+With the removal of `extends`, it's now up to the users to inherit from other configs. That can easily be done in the following manner:
+
+```js
+const standardConfig = require("eslint-config-standard");
+
+standardConfig.rules.semi = ["error", "always"];
+
+exports.config = standardConfig;
+```
+
+This config extends `eslint-config-standard` by first importing and then making modifications before exporting the modified config.
+
+#### Overriding Configuration Based on File Patterns
+
+Whereas `.eslintrc` had an `overrides` key that made a hierarchical structure, the `.eslint.js` file does not have any such hierarchy. Instead, users can return an array of configs that should be used. For example, consider this `.eslintrc` config:
+
+```yaml
+plugins: ["react"]
+rules:
+   react/jsx-uses-react: error
+    semi: error
+
+overrides:
+    - files: "*.md"
+      plugins: ["markdown"],
+      processor: "markdown/markdown"
+```
+
+This can be written in `.eslint.js` as an array of two configs:
+
+```js
+exports.config = [
+    {
+        files: "*.js",
+        ruledefs: {
+            react: require("eslint-plugin-react").rules,
+        },
+        rules: {
+            "react/jsx-uses-react": "error",
+            semi: "error"
+        }
+    },
+    {
+        files: "*.md",
+        processor: require("eslint-plugin-markdown").processors.markdown
+    }
+];
+```
+
+When ESLint uses this config, it will check each `files` pattern to determine which config applies. In this way, returning an array acts exactly the same as the array in `overrides`, with the exception that no configuration options are inherited.
+
+#### Replacing `.eslintignore`
+
+Because there is only one `.eslint.js` file to consider, ESLint doesn't have to first search directories to determine its location. That allows `.eslint.js` to specify files to ignore directly instead of relying on `.eslintignore`. For backwards compatibility, users could create a config like this:
```

Comment:
I'm not sure that's a correct assumption. Just because each config file basically acts as if `root: true` is specified, doesn't mean that users will choose not to create a separate config file in their `tests` directory, for example. They might even inherit manually from the higher config file by requiring it.

Reply:
That's true, though I would expect they would use a `files` pattern for that purpose rather than creating a separate file. I think my general use case assumption still makes sense here.